// @bbob/plugin-helper/es/index.js
"use strict";

exports.__esModule = true;
exports.isEOL = exports.isStringNode = exports.isTagNode = exports.getUniqAttr = exports.getNodeLength = exports.appendToNode = exports.attrValue = exports.attrsToString = void 0;

var _char = require("./char");

var isTagNode = el => typeof el === 'object' && !!el.tag;

exports.isTagNode = isTagNode;

var isStringNode = el => typeof el === 'string';

exports.isStringNode = isStringNode;

var isEOL = el => el === _char.N;

exports.isEOL = isEOL;

var getNodeLength = node => {
  if (isTagNode(node)) {
    return node.content.reduce((count, contentNode) => count + getNodeLength(contentNode), 0);
  }

  if (isStringNode(node)) {
    return node.length;
  }

  return 0;
};
/**
 * Appends value to Tag Node
 * @param {TagNode} node
 * @param value
 */


exports.getNodeLength = getNodeLength;

var appendToNode = (node, value) => {
  node.content.push(value);
};
/**
 * Replaces " to &qquot;
 * @param {String} value
 */


exports.appendToNode = appendToNode;

var escapeQuote = value => value.replace(/"/g, '&quot;');
/**
 * Acept name and value and return valid html5 attribute string
 * @param {String} name
 * @param {String} value
 * @return {string}
 */


var attrValue = (name, value) => {
  var type = typeof value;
  var types = {
    boolean: () => value ? "" + name : '',
    number: () => name + "=\"" + value + "\"",
    string: () => name + "=\"" + escapeQuote(value) + "\"",
    object: () => name + "=\"" + escapeQuote(JSON.stringify(value)) + "\""
  };
  return types[type] ? types[type]() : '';
};
/**
 * Transforms attrs to html params string
 * @param values
 */


exports.attrValue = attrValue;

var attrsToString = values => {
  // To avoid some malformed attributes
  if (typeof values === 'undefined') {
    return '';
  }

  return Object.keys(values).reduce((arr, key) => [...arr, attrValue(key, values[key])], ['']).join(' ');
};
/**
 * Gets value from
 * @example
 * getUniqAttr({ 'foo': true, 'bar': bar' }) => 'bar'
 * @param attrs
 * @returns {string}
 */


exports.attrsToString = attrsToString;

var getUniqAttr = attrs => Object.keys(attrs).reduce((res, key) => attrs[key] === key ? attrs[key] : null, null);

exports.getUniqAttr = getUniqAttr;
